clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -6389,10 +6389,6 @@ static bool checkIntelFPGARegisterAttrCompatibility(Sema &S, Decl *D,
     if (!MA->isImplicit() &&
         checkAttrMutualExclusion<IntelFPGAMemoryAttr>(S, D, Attr))
       InCompat = true;
-  if (auto *NBA = D->getAttr<IntelFPGANumBanksAttr>())
```

Comment:
This change has me a bit worried (along with the new `MutualExclusions` definition in Attr.td) -- it used to be we'd only look for a mutual exclusion when the `numbanks` attribute was not implicit. Now we're checking it always, but we still create a numbanks implicitly (see `Sema::AddIntelFPGABankBitsAttr()`).

Reply:
Sorry I just noticed your message. 
>>Now we're checking it always, but we still create a numbanks implicitly (see Sema::AddIntelFPGABankBitsAttr()).

Yes, correct

Reply:
I checked AST for this test case without refactoring patch:
//expected-error@+2 1{{'fpga_register' and 'bank_bits' attributes are not compatible}}
  [[intel::bank_bits(2, 3)]]
  [[intel::fpga_register]]
  //expected-note@-2 1{{conflicting attribute is here}}
  unsigned int bb_reg[4];


```
 |-DeclStmt 0x5592c9ea9680 <line:558:3, line:561:25>
|   | `-VarDecl 0x5592c9ea94d0 <col:3, col:24> col:16 bb_reg 'unsigned int [4]'
|   |   |-IntelFPGANumBanksAttr 0x5592c9ea95b0 <<invalid sloc>> Implicit
|   |   | `-IntegerLiteral 0x5592c9ea9590 <<invalid sloc>> 'int' 4
|   |   |-IntelFPGAMemoryAttr 0x5592c9ea9610 <<invalid sloc>> Implicit memory Default
|   |   `-IntelFPGABankBitsAttr 0x5592c9ea9638 <line:558:5, col:26>
|   |     |-ConstantExpr 0x5592c9ea9550 <col:22> 'int'
|   |     | |-value: Int 2
|   |     | `-IntegerLiteral 0x5592c9ea9428 <col:22> 'int' 2
|   |     `-ConstantExpr 0x5592c9ea9570 <col:25> 'int'
|   |       |-value: Int 3
|   |       `-IntegerLiteral 0x5592c9ea9448 <col:25> 'int' 3

```

The same happens with current factoring patch.

Reply:
Okay, thank you for checking!